### PR TITLE
add some internal caches

### PR DESCRIPTION
**Pull request by ltalirz**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation
- [x] Developer tools
- [x] Refactoring
- [ ] Test

## Actions (for code changes)

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by ltalirz** [comment]
* switch from prospector to pylint
 * add pre-commit extra
 * switch to isolated hooks for black and isort
 * add caches for structure graph, connected sites and coordination number
 * use ordered dict for results
 * allow using multiple CLI options

**Comment by ltalirz** [comment]
For a test structure with 1300 atoms, these changes do reduce the time spent in `get_mof_descriptors` from 58s to 28s, but in the end the problem is that there is no internal caching inside pymatgen - e.g. many calls end up in the internal `get_nn_info` method (which is also not written in a particularly efficient way).

I'm wondering: would there be faster implementations that one can use?

**Comment by kjappelbaum** [comment]
I was at the same question for the oximachine at some point and they also had an open issue for that (will look it up and add here). Might be worth to go back to it.

File: .pylintrc [hunk 1]
```diff
@@ -3,4 +3,4 @@ disable = C0330, C0326, locally-disabled, fixme
 
 
 [format]
-max-line-length = 88
```
Comment (kjappelbaum) [inline .pylintrc:6]: We had 88 before since we use black here (https://black.readthedocs.io/en/stable/compatible_configs.html)

**Review by kjappelbaum — APPROVED** [review]
Many thanks, @ltalirz 

Greatly improved the codebase. At some point I'll still need to refactor more (as it is not great that one currently has to define each new check in three different files)

File: .pylintrc (hunk 1 above)
Comment (ltalirz) [reply .pylintrc]: makes sense - I just changed it because I noticed that it failed on a line I didn't touch

I've fixed it now + added pre-commit to CI

**Pull request merged**